Cache m_cached_finished_ibd where SetTip is called.

pstratem opened the pull request:
As IsInitialBlockDownload latches to false only once the Tip is sufficiently
advanced there is no need to check the Tip everytime IsIBD is called.

By caching this in advance we can avoid extra work and more importantly a lock.

stickies-v reviewed (commented):
Conceptually not a bad idea to cache and lock less, but imo this makes the code more brittle (and harder to understand), e.g. if any tip updates happen without the cache being updated separately.

Do you have any data as to the actual performance improvements from this PR?

furszy reviewed (commented):
I think something like this, if properly implemented (I haven't thought much about the code yet), would reduce the GUI freezes during IBD in a noticeable manner.

pstratem commented:
> Conceptually not a bad idea to cache and lock less, but imo this makes the code more brittle (and harder to understand), e.g. if any tip updates happen without the cache being updated separately.
> 
> Do you have any data as to the actual performance improvements from this PR?

I'm (very) open to suggestions on how to make the caching call more robust. (Indeed I expected some.)

There's no performance improvement from this PR, it's the first in a series of proposed changes I'll be making to remove locking where it's not necessary, with the end goal being some form of concurrency being possible in message processing.

pstratem commented:
> I think something like this, if properly implemented (I haven't thought much about the code yet), would reduce the GUI freezes during IBD in a noticeable manner.

I hadn't even considered that, but certainly that's a possible direct improvement.

luke-jr reviewed (changes requested):
Concept ACK, but I'm not convinced this implementation is safe as-is. If we want to maintain the current behaviour, it's not sufficient to update only when the tip changes. We also need to re-check when importing/reindexing completes, and schedule an update timer if max_tip_age is the final cause of not exiting IBD.

pstratem commented:
> Concept ACK, but I'm not convinced this implementation is safe as-is. If we want to maintain the current behaviour, it's not sufficient to update only when the tip changes. We also need to re-check when importing/reindexing completes, and schedule an update timer if max_tip_age is the final cause of not exiting IBD.

This made me revisit the function and consider what we're trying to achieve.

The function is only interesting when it can latch to the IBD finished state.

That's only possible when all four conditions are met, which can only happen when the tip is updated.

The final time based condition can only change when the tip changes as it gets further away with time, not closer.

mzumsande commented on src/validation.cpp:
```diff
@@ -3089,6 +3079,7 @@ bool Chainstate::DisconnectTip(BlockValidationState& state, DisconnectedBlockTra
     }
 
     m_chain.SetTip(*pindexDelete->pprev);
+    m_chainman.CacheIsInitialBlockDownload();
```
First I thought this was not necessary because disconnecting a block shouldn't usually get you out of IBD, but it guess there are edge cases (starting up, with the old tip having a lower timestamp than it's parent block) where this could lead to get us out of IBD?!

mzumsande commented on src/validation.cpp:
```diff
@@ -2010,36 +2010,26 @@ void Chainstate::InitCoinsCache(size_t cache_size_bytes)
+    if (m_blockman.LoadingBlocks()) return;
+
+    {
+        AssertLockHeld(cs_main);
```
since the function is annotated with `EXCLUSIVE_LOCKS_REQUIRED(cs_main)` anyway, why not put it to the beginning of the function, as it is done in most other places?

mzumsande reviewed (commented):
> The function is only interesting when it can latch to the IBD finished state.
> 
> That's only possible when all four conditions are met, which can only happen when the tip is updated.
> 
> The final time based condition can only change when the tip changes as it gets further away with time, not closer.

I think that @luke-jr is right. If we reindex, we set `m_importing` to true in `ImportBlocks`, so any blocks we connect there can never result in getting out of IBD due to the `m_blockman.LoadingBlocks()` [early return](https://github.com/pstratem/bitcoin/blob/2e3fefbaaa01f42d836bce8dcad5cf3b80971f4e/src/validation.cpp#L2022). 
Therefore we need a call to `CacheIsInitialBlockDownload()` after `ImportingNow` goes out of scope in `ImportBlocks()`.

Commit:
Add m_cached_chaintip_recent and method to update it UpdateCachedChaintipRecent.

On systems with sane clocks the chain tip checks can only change when the tip
changes. The gap between the chain tip and the current time only grows.

Commit:
Call UpdateCachedChaintipRecent when SetTip is moving the active chain forward.

Commit:
Use m_cached_chaintip_recent in IsIBD removing the cs_main lock.

pstratem commented:
Ok I thought about it and it just wasn't obviously correct enough.

So I've rewritten into three commits to be simpler.

l0rinc commented on src/test/util/validation.cpp:
```diff
@@ -21,6 +21,7 @@ void TestChainstateManager::DisableNextWrite()
 void TestChainstateManager::ResetIbd()
 {
     m_cached_finished_ibd = false;
+    m_cached_chaintip_recent = false;
```
I find the existence of his whole method very hacky, we're testing something that cannot happen in reality so if the test passes or fails after this, it won't increase my confidence in the product.
But if you insist on updating it, we should likely update `JumpOutOfIbd` as well for symmetry.

l0rinc commented on src/validation.h:
```diff
@@ -1142,6 +1143,7 @@ class ChainstateManager
 
     /** Check whether we are doing an initial block download (synchronizing from disk or network) */
     bool IsInitialBlockDownload() const;
+    void UpdateCachedChaintipRecent() EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```
should be const and needs some comment (and I'd specialize it to just return the value instead of mutating the state, we can do that in the `SetTip` method instead)
```suggestion
    /** Check whether the active chain tip exists, has enough work, and is recent. */
    bool IsTipRecent() const EXCLUSIVE_LOCKS_REQUIRED(cs_main);
```

l0rinc commented on src/validation.cpp:
```diff
@@ -2040,6 +2040,17 @@ bool ChainstateManager::IsInitialBlockDownload() const
+    if (chain.Tip()->nChainWork < MinimumChainWork()) return;
+    if (chain.Tip()->Time() < Now<NodeSeconds>() - m_options.max_tip_age) return;
+
+    m_cached_chaintip_recent = true;
```
shouldn't we guard this method with this being false?

l0rinc commented on src/validation.h:
```diff
@@ -1029,6 +1029,7 @@ class ChainstateManager
      * const, which latches this for caching purposes.
      */
     mutable std::atomic<bool> m_cached_finished_ibd{false};
+    mutable std::atomic<bool> m_cached_chaintip_recent{false};
```
no need to mention `chain` here and we can use `std::atomic_bool` instead and should add some description to it

l0rinc reviewed (changes requested):
Concept ACK, makes sense to push the burden to the writer instead of the reader.
But we need to restructure it slightly so that it tells a story of what we're extracting, delegating and caching exactly.

I have implemented an example in https://github.com/l0rinc/bitcoin/pull/60 (prototype, may not pass all tests yet).

l0rinc commented:
@pstratem, are you still working on this or would you like me to take over?
